File: Python/A_Project/A_Tic_Tac_Toe/PlayerStrategies/ConcreteStrategy/human_player_strategy.py
from PlayerStrategies.player_strategy import PlayerStrategy
from Utility.position import Position


class HumanPlayerStrategy(PlayerStrategy):
    def __init__(self, player_name: str):
        self.player_name = player_name

    def make_move(self, board):
        while True:
            prompt = (
                f"{self.player_name}, enter your move (row [0-{board.rows - 1}] "
                f"and column [0-{board.cols - 1}]): "
            )
            raw = input(prompt)
            parts = raw.strip().split()
            if len(parts) < 2:
                print("Invalid input. Please enter numbers for row and column.")
                continue
            try:
                row = int(parts[0])
                col = int(parts[1])
            except ValueError:
                print("Invalid input. Please enter numbers for row and column.")
                continue

            move = Position(row, col)
            if board.is_valid_move(move):
                return move

            print("This move is not valid. Try again.")

File: Ruby/A_Project/A_Tic_Tac_Toe/PlayerStrategies/ConcreteStrategy/human_player_strategy.rb
# Used Strategy Design Pattern

require_relative '../player_strategy'
require_relative '../../Utility/position'

module PlayerStrategies
  module ConcreteStrategy
    class HumanPlayerStrategy
      include PlayerStrategy

      def initialize(player_name)
        @player_name = player_name
      end

      def make_move(board)
        loop do
          print "#{@player_name}, enter your move (row [0-2] and column [0-2]): "

          begin
            input = gets.chomp.split
            raise ArgumentError unless input.length == 2

            row = Integer(input[0])
            col = Integer(input[1])
            move = Utility::Position.new(row, col)

            return move if board.valid_move?(move)

            puts "This move is not valid. Try again."
          rescue ArgumentError
            puts "Invalid input. Please enter numbers for row and column."
          end
        end
      end
    end
  end
end
